modules/web/middleware/cookie.go

```diff
@@ -32,15 +32,24 @@ func GetSiteCookie(req *http.Request, name string) string {
 	return val
 }
 
+// GetCookieSecure returns whether the "Secure" attribute on a cookie should be set
+func GetCookieSecure(req *http.Request) bool {
+	forwardedProto := strings.ToLower(req.Header.Get("x-forwarded-proto"))
+	if forwardedProto != "" {
+		return forwardedProto == "https" || forwardedProto == "wss"
+	}
+	return req.TLS != nil
+}
```

Comment:
Sorry but it's not right.

Many users do not pass these header when they use Gitea behind a reverse proxy.

Reply:
Maybe we can just parse `setting.AppURL`.

Reply:
> Maybe we can just parse `setting.AppURL`.

Yes. And:

1. Keep the COOKIE_SECURE option, in case some people really would like to run Gitea on different domains
    * COOKIE_SECURE should be default to true if AppURL is "https"
2. Show a frontend warning if the "http" doesn't match COOKIE_SECURE=true, to tell users that "you might not be able to login"

Reply:
> Many users might not pass these header when they use Gitea behind a reverse proxy.

At least for nginx, we have been documenting for a long time to configure that header.

Reply:
Yes, only "At least". None of others are documented.

And even there is a document, many users do not follow that, they just like to write the config by their experiences.

Reply:
I would call it out in changelog and be done with it. It's the right way to check if connection is secure imho.

Reply:
Thought I do see parsing `AppURL` as the more forgiving option. In an application that does not have such a setting, this method is the only one that can work, but given that we have `AppURL` we might as well use it.

What does not work with `AppURL` is to host the application on multiple domains with mixed http/https but I guess that is a seldomly used configuration.

Reply:
I prefer the solution in comment https://github.com/go-gitea/gitea/pull/26953#discussion_r1317993450 .

According to MDN, the `X-Forwarded-Proto` is not standard and there are many alternatives (`Front-End-Https` / `X-Forwarded-Protocol` /  `X-Forwarded-Ssl` /  `X-Url-Scheme`). IMO using the strong assumption with `X-Forwarded-Proto` might cause many users unable to login and fail silently. Gitea should be able to be easy to setup and use out-of-box.

----

My suggestion also works well with "host the application on multiple domains and/or mixed http/https", the site admin could set COOKIE_SECURE=false in their config, and redirect untrusted HTTP requests to HTTPS, still no security risk.

Reply:
So in summary: Keep `COOKIE_SECURE` but the AppUrl is https, switch it to true? I'm not sure such "magic" is good to do. Maybe I should just scrap the entire PR.